modules/ads_lib/src/main/java/com/google/api/ads/adwords/lib/utils/BatchJobHelperUtility.java

```diff
@@ -0,0 +1,135 @@
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.api.ads.adwords.lib.utils;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Supplier;
+import com.google.common.io.ByteSource;
+import com.google.inject.Inject;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.StringWriter;
+import java.net.URL;
+import javax.xml.stream.XMLEventFactory;
+import javax.xml.stream.XMLEventReader;
+import javax.xml.stream.XMLEventWriter;
+import javax.xml.stream.XMLInputFactory;
+import javax.xml.stream.XMLOutputFactory;
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.events.EndElement;
+import javax.xml.stream.events.StartElement;
+import javax.xml.stream.events.XMLEvent;
+import javax.xml.transform.OutputKeys;
+import javax.xml.transform.Source;
+import javax.xml.transform.Transformer;
+import javax.xml.transform.TransformerException;
+import javax.xml.transform.stream.StreamResult;
+import javax.xml.transform.stream.StreamSource;
+
+/** The utility class to be used in {@code BatchJobHelperInterface} */
+public class BatchJobHelperUtility {
+
+  private static final String TAG_MUTATE_RESPONSE = "mutateResponse";
+  private static final String TAG_RVAL = "rval";
+
+  private final Supplier<Transformer> transformerSupplier;
+  private final XMLInputFactory xmlInputFactory;
+  private final XMLOutputFactory xmlOutputFactory;
+  private final XMLEventFactory xmlEventFactory;
+
+  @Inject
+  BatchJobHelperUtility(
+      Supplier<Transformer> transformerSupplier,
+      XMLInputFactory xmlInputFactory,
+      XMLOutputFactory xmlOutputFactory,
+      XMLEventFactory xmlEventFactory) {
+    this.transformerSupplier = transformerSupplier;
+    this.xmlInputFactory = xmlInputFactory;
+    this.xmlOutputFactory = xmlOutputFactory;
+    this.xmlEventFactory = xmlEventFactory;
+  }
+
+  /**
+   * Returns a {@code ByteSource} object that wraps the batch job response into chunks with
+   * pagination, which will be parsed by Axis/JAX-WS.
+   */
+  public ByteSource buildWrappedByteSource(URL url, int startIndex, int numberResults)
+      throws IOException, XMLStreamException, TransformerException {
+    Preconditions.checkArgument(startIndex >= 0, "startIndex must not be negative");
+    Preconditions.checkArgument(numberResults > 0, "numberResults must be positive");
+
+    StringWriter stringWriter = new StringWriter();
+
+    XMLEventReader xmlEventReader = xmlInputFactory.createXMLEventReader(url.openStream());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *openStream*. Currently, there are execution paths that do not contain closure statements, for example, when XMLInputFactory.createXMLEventReader throws an exception. Either a) close the object returned by *openStream()* in a try-finally block or b) close the resource by declaring the object returned by *openStream()* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).